bdai_ros2_wrappers/bdai_ros2_wrappers/executors.py

```diff
@@ -0,0 +1,414 @@
+        done, not_done = concurrent.futures.wait(futures, timeout=timeout)
+        return len(not_done) == 0
+
+    def _cleanup_after(self, work: "AutoScalingThreadPool.Work") -> bool:
+        complete = True
+        with self._submit_lock:
+            self._runlists[work.fn].remove(work)
+            if work.fn in self._waitqueues and self._waitqueues[work.fn]:  # continue with pending work
+                work = self._waitqueues[work.fn].popleft()
+                while work.cancelled() and self._waitqueues[work.fn]:
+                    work.notify_cancelation()
+                    work = self._waitqueues[work.fn].popleft()
+                if not self._waitqueues[work.fn]:
+                    del self._waitqueues[work.fn]
+                if not work.cancelled():
+                    self._runlists[work.fn].append(work)
+                    self._runqueue.put(work)  # actually submit work
+                    complete = False
+                else:
+                    work.notify_cancelation()
+            if not self._runlists[work.fn]:
+                del self._runlists[work.fn]
+        return complete
+
+    @classmethod
+    def _do_work(cls, executor_weakref: weakref.ref, logger: logging.Logger, stop_on_timeout: bool = True) -> None:
+        try:
+            work: typing.Optional[AutoScalingThreadPool.Work] = None
+            while not cls._interpreter_shutdown:
+                executor: typing.Optional[AutoScalingThreadPool] = executor_weakref()
+                if executor is None:
+                    break
+
+                runqueue = executor._runqueue
+                if executor._shutdown:
+                    runqueue.put(None)
+                    break
+
+                runslots = executor._runslots
+                if work is not None:
+                    if executor._cleanup_after(work):
+                        runslots.release()
+                    work = None
+                else:
+                    runslots.release()
+
+                timeout = executor._max_idle_time
+                del executor  # drop reference
+
+                try:
+                    work = runqueue.get(block=True, timeout=timeout)
+                except queue.Empty:
+                    if stop_on_timeout:  # non-stable worker
+                        if runslots.acquire(blocking=False):
+                            break
+                    continue
+
+                if work is not None:
+                    work.execute()
+        except BaseException:
+            import textwrap
+            import traceback
+
+            logger.error("Worker threw an exception: \n%s", textwrap.indent(traceback.format_exc(), "  "))
+
+    def _do_submit(self, work: "AutoScalingThreadPool.Work") -> None:
+        self._workers = weakref.WeakSet([w for w in self._workers if w.is_alive()])
+        while not self._runslots.acquire(timeout=self._submission_patience):
```

Comment:
I'm a bit confused about this part. If the semaphore is full, this loop spins up another thread to run the new function right? Does the counter on the semaphore or number of things that can access the semaphore get updated accordingly?

Reply:
Yes, the semaphore ie. the runslot count gets increased by workers. Submitting work 
decreases the count, workers ready to work increase it. I added documentation to clarify this.